python/pyarrow/tests/test_fs.py

```diff
@@ -0,0 +1,324 @@
+        id='LocalFileSystem'
+    ),
+    pytest.param(
+        lambda tmp: SubTreeFileSystem(tmp, LocalFileSystem()),
+        id='SubTreeFileSystem(LocalFileSystem)'
+    )
+])
+def fs(request, tempdir):
+    return request.param(tempdir.as_posix())
+
+
+@pytest.fixture(params=[
+    pytest.param(pathlib.PurePosixPath, id='Path'),
+    pytest.param(str, id='str')
+])
+def testpath(request, fs, tempdir):
+    # we always use the tempdir for reading and writing test artifacts, but
+    # if the filesystem is wrapped in a SubTreeFileSystem then we don't need
+    # to prepend the path with the tempdir, we also test the API with both
+    # pathlib.Path objects and plain python strings
+    def convert(path):
+        if isinstance(fs, SubTreeFileSystem):
+            path = pathlib.PurePosixPath(path)
+        else:
+            path = tempdir / path
+        # convert to abstract, slash separated paths with as_posix
+        path = path.as_posix()
+        # return with the corrent
+        return request.param(path)
+    return convert
+
+
+def test_cannot_instantiate_base_filesystem():
+    with pytest.raises(TypeError):
+        FileSystem()
+
+
+def test_non_path_like_input_raises(fs):
+    class Path:
+        pass
+
+    invalid_paths = [1, 1.1, Path(), tuple(), {}, [], lambda: 1]
+    for path in invalid_paths:
+        with pytest.raises(TypeError):
+            fs.create_dir(path)
+
+
+def test_get_target_stats(fs, tempdir, testpath):
+    aaa, aaa_ = testpath('a/aa/aaa'), tempdir / 'a' / 'aa' / 'aaa'
+    bb, bb_ = testpath('a/bb'), tempdir / 'a' / 'bb'
+    c, c_ = testpath('c.txt'), tempdir / 'c.txt'
+
+    aaa_.mkdir(parents=True)
+    bb_.touch()
+    c_.write_bytes(b'test')
+
+    def mtime_almost_equal(fs_dt, pathlib_ts):
+        # arrow's filesystem implementation truncates mtime to microsends
+        # resolution whereas pathlib rounds
+        pathlib_dt = datetime.utcfromtimestamp(pathlib_ts)
+        difference = (fs_dt - pathlib_dt).total_seconds()
+        return abs(difference) <= 10**-6
+
+    aaa_stat, bb_stat, c_stat = fs.get_target_stats([aaa, bb, c])
+
+    assert aaa_stat.path == str(aaa)
+    assert aaa_stat.base_name == 'aaa'
+    assert aaa_stat.extension == ''
+    assert aaa_stat.type == FileType.Directory
+    assert mtime_almost_equal(aaa_stat.mtime, aaa_.stat().st_mtime)
+    with pytest.raises(ValueError):
+        aaa_stat.size
+
+    assert bb_stat.path == str(bb)
+    assert bb_stat.base_name == 'bb'
+    assert bb_stat.extension == ''
+    assert bb_stat.type == FileType.File
+    assert bb_stat.size == 0
+    assert mtime_almost_equal(bb_stat.mtime, bb_.stat().st_mtime)
+
+    assert c_stat.path == str(c)
+    assert c_stat.base_name == 'c.txt'
+    assert c_stat.extension == 'txt'
+    assert c_stat.type == FileType.File
+    assert c_stat.size == 4
+    assert mtime_almost_equal(c_stat.mtime, c_.stat().st_mtime)
+
+
+@pytest.mark.parametrize('base_dir', ['.', '..'])
```

Comment:
I don't understand why you're testing with ".." here. It doesn't seem useful (and also seems responsible for the Windows failures).